server/src/main/java/org/opensearch/action/search/CreatePitController.java

```diff
@@ -0,0 +1,253 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.StepListener;
+import org.opensearch.action.support.GroupedActionListener;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Strings;
+import org.opensearch.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.search.SearchPhaseResult;
+import org.opensearch.search.SearchShardTarget;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.Transport;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+import static org.opensearch.common.unit.TimeValue.timeValueSeconds;
+
+/**
+ * Controller for creating PIT reader context
+ * Phase 1 of create PIT request : Create PIT reader contexts in the associated shards with a temporary keep alive
+ * Phase 2 of create PIT : Update PIT reader context with PIT ID and keep alive from request and
+ * fail user request if any of the updates in this phase are failed - we clean up PITs in case of such failures.
+ * This two phase approach is used to save PIT ID as part of context which is later used for other use cases like list PIT etc.
+ */
+public class CreatePitController {
+    private final SearchTransportService searchTransportService;
+    private final ClusterService clusterService;
+    private final TransportSearchAction transportSearchAction;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private final Task task;
+    private final ActionListener<CreatePitResponse> listener;
+    private final CreatePitRequest request;
+    private static final Logger logger = LogManager.getLogger(CreatePitController.class);
+    public static final Setting<TimeValue> PIT_INIT_KEEP_ALIVE = Setting.positiveTimeSetting(
+        "pit.init.keep_alive",
+        timeValueSeconds(30),
+        Setting.Property.NodeScope
+    );
+
+    public CreatePitController(
+        CreatePitRequest request,
+        SearchTransportService searchTransportService,
+        ClusterService clusterService,
+        TransportSearchAction transportSearchAction,
+        NamedWriteableRegistry namedWriteableRegistry,
+        Task task,
+        ActionListener<CreatePitResponse> listener
+    ) {
```

Comment:
Lets create a singleton `CreatePitController ` and remove the `CreatePitRequest ` from the constructor, it should be a part of `execute` request part